<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of parseHPArna</title>
  <meta name="keywords" content="parseHPArna">
  <meta name="description" content="parseHPA">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">hpa</a> &gt; parseHPArna.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for hpa&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>parseHPArna
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>parseHPA</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function arrayData=parseHPArna(fileName, version) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> parseHPA
   Parses a database dump of the Human Protein Atlas (HPA) RNA-Seq data.

   Input:
   fileName            tab-separated database dump of HPA RNA data. For
                       details regarding the format, see
                       http://www.proteinatlas.org/about/download.
   version             version of HPA [optional, default=19]


   Output:
   arrayData
       genes               cell array with the unique ensemble gene IDs
       geneNames           cell array with the gene names (gene abbrevs)
       tissues             cell array with the tissue names
       levels              matrix of gene expression levels (TPM), where
                           rows correspond to genes, and columns
                           correspond to tissues

   Usage: arrayData=parseHPArna(fileName,version)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function arrayData=parseHPArna(fileName, version)</a>
0002 <span class="comment">% parseHPA</span>
0003 <span class="comment">%   Parses a database dump of the Human Protein Atlas (HPA) RNA-Seq data.</span>
0004 <span class="comment">%</span>
0005 <span class="comment">%   Input:</span>
0006 <span class="comment">%   fileName            tab-separated database dump of HPA RNA data. For</span>
0007 <span class="comment">%                       details regarding the format, see</span>
0008 <span class="comment">%                       http://www.proteinatlas.org/about/download.</span>
0009 <span class="comment">%   version             version of HPA [optional, default=19]</span>
0010 <span class="comment">%</span>
0011 <span class="comment">%</span>
0012 <span class="comment">%   Output:</span>
0013 <span class="comment">%   arrayData</span>
0014 <span class="comment">%       genes               cell array with the unique ensemble gene IDs</span>
0015 <span class="comment">%       geneNames           cell array with the gene names (gene abbrevs)</span>
0016 <span class="comment">%       tissues             cell array with the tissue names</span>
0017 <span class="comment">%       levels              matrix of gene expression levels (TPM), where</span>
0018 <span class="comment">%                           rows correspond to genes, and columns</span>
0019 <span class="comment">%                           correspond to tissues</span>
0020 <span class="comment">%</span>
0021 <span class="comment">%   Usage: arrayData=parseHPArna(fileName,version)</span>
0022 
0023 <span class="keyword">if</span> nargin&lt;2
0024     <span class="comment">%Change this and add code for more versions when the current HPA</span>
0025     <span class="comment">%version is increased and the format is changed</span>
0026     version=19;
0027 <span class="keyword">end</span>
0028 
0029 <span class="keyword">if</span> ~(exist(fileName,<span class="string">'file'</span>)==2)
0030     error(<span class="string">'HPA file %s cannot be found'</span>, string(fileName));
0031 <span class="keyword">end</span>
0032 
0033 <span class="comment">%NOTE! This function assumes that the first 4 columns contain (in order):</span>
0034 <span class="comment">% (1) gene ensembl ID, (2) gene abbrev, (3) tissue name,  (4) TPM value</span>
0035 <span class="keyword">if</span> (version == 19)
0036     headers={<span class="string">'Gene'</span> <span class="string">'Gene name'</span> <span class="string">'Tissue'</span> <span class="string">'TPM'</span> <span class="string">'pTPM'</span> <span class="string">'NX'</span>};
0037 <span class="keyword">elseif</span> (version == 18)
0038     headers={<span class="string">'Gene'</span> <span class="string">'Gene name'</span> <span class="string">'Sample'</span> <span class="string">'Value'</span> <span class="string">'Unit'</span>};
0039 <span class="keyword">else</span>
0040     error(<span class="string">'Only HPA versions 18 and 19 are currently supported.'</span>);
0041 <span class="keyword">end</span>
0042 
0043 <span class="comment">%extract data from file</span>
0044 formatSpec = strip(repmat(<span class="string">'%q '</span>, 1, numel(headers)));
0045 fid=fopen(fileName, <span class="string">'r'</span>);
0046 hpa=textscan(fid, formatSpec, <span class="string">'Delimiter'</span>, <span class="string">'\t'</span>);
0047 fclose(fid);
0048 
0049 <span class="comment">%Go through and see if the headers match what was expected</span>
0050 <span class="keyword">for</span> i=1:numel(headers)
0051     <span class="keyword">if</span> ~strcmpi(headers(i),hpa{i}(1))
0052         EM=[<span class="string">'Could not find the header &quot;'</span> headers{i} <span class="string">'&quot;. Make sure that the input file matches the format specified at http://www.proteinatlas.org/about/download'</span>];
0053         dispEM(EM);
0054     <span class="keyword">end</span>
0055     <span class="comment">%Remove the header line here</span>
0056     hpa{i}(1)=[];
0057 <span class="keyword">end</span>
0058 
0059 <span class="comment">%Get unique gene IDs and tissue names</span>
0060 [arrayData.genes, P, I] = unique(hpa{1});
0061 arrayData.geneNames = hpa{2}(P);  <span class="comment">% retrieve corresponding gene names</span>
0062 [arrayData.tissues, ~, J] = unique(hpa{3});
0063 
0064 <span class="comment">%Now extract the gene levels and organize into matrix</span>
0065 arrayData.levels = NaN(max(I),max(J));
0066 linearInd = sub2ind(size(arrayData.levels),I,J);
0067 arrayData.levels(linearInd) = str2double(hpa{4});
0068     
0069</pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>