<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getDiamond</title>
  <meta name="keywords" content="getDiamond">
  <meta name="description" content="getDiamond">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getDiamond.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getDiamond
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getDiamond</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function blastStructure=getDiamond(organismID,fastaFile,modelIDs,refFastaFiles) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getDiamond
   Uses DIAMOND to performs a bidirectional BLASTP between the organism
   of interest and a set of template organisms.

   Input:
   organismID      the id of the organism of interest. This should also
                   match with the id supplied to getModelFromHomology
   fastaFile       a FASTA file with the protein sequences for the
                   organism of interest
   modelIDs        a cell array of model ids. These must match the
                   &quot;model.id&quot; fields in the &quot;models&quot; structure if the
                   output is to be used with getModelFromHomology
   refFastaFiles   a cell array with the paths to the corresponding FASTA
                   files

   Output:
   blastStructure  structure containing the bidirectional homology
                   measurements which are used by getModelFromHomology

   NOTE: This function calls DIAMOND to perform a bidirectional homology
   test between the organism of interest and a set of other organisms
   using the '--more-sensitive' setting from DIAMOND. For the most
   sensitive results, the use of getBlast() is adviced, however,
   getDiamond() is a fast alternative (&gt;15x faster). The blastStructure
   generated is in the same format as those obtained from getBlast().

   Usage: blastStructure=getDiamond(organismID,fastaFile,modelIDs,...
           refFastaFiles)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function blastStructure=getDiamond(organismID,fastaFile,modelIDs,refFastaFiles)</a>
0002 <span class="comment">% getDiamond</span>
0003 <span class="comment">%   Uses DIAMOND to performs a bidirectional BLASTP between the organism</span>
0004 <span class="comment">%   of interest and a set of template organisms.</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   Input:</span>
0007 <span class="comment">%   organismID      the id of the organism of interest. This should also</span>
0008 <span class="comment">%                   match with the id supplied to getModelFromHomology</span>
0009 <span class="comment">%   fastaFile       a FASTA file with the protein sequences for the</span>
0010 <span class="comment">%                   organism of interest</span>
0011 <span class="comment">%   modelIDs        a cell array of model ids. These must match the</span>
0012 <span class="comment">%                   &quot;model.id&quot; fields in the &quot;models&quot; structure if the</span>
0013 <span class="comment">%                   output is to be used with getModelFromHomology</span>
0014 <span class="comment">%   refFastaFiles   a cell array with the paths to the corresponding FASTA</span>
0015 <span class="comment">%                   files</span>
0016 <span class="comment">%</span>
0017 <span class="comment">%   Output:</span>
0018 <span class="comment">%   blastStructure  structure containing the bidirectional homology</span>
0019 <span class="comment">%                   measurements which are used by getModelFromHomology</span>
0020 <span class="comment">%</span>
0021 <span class="comment">%   NOTE: This function calls DIAMOND to perform a bidirectional homology</span>
0022 <span class="comment">%   test between the organism of interest and a set of other organisms</span>
0023 <span class="comment">%   using the '--more-sensitive' setting from DIAMOND. For the most</span>
0024 <span class="comment">%   sensitive results, the use of getBlast() is adviced, however,</span>
0025 <span class="comment">%   getDiamond() is a fast alternative (&gt;15x faster). The blastStructure</span>
0026 <span class="comment">%   generated is in the same format as those obtained from getBlast().</span>
0027 <span class="comment">%</span>
0028 <span class="comment">%   Usage: blastStructure=getDiamond(organismID,fastaFile,modelIDs,...</span>
0029 <span class="comment">%           refFastaFiles)</span>
0030 
0031 <span class="comment">%Everything should be cell arrays</span>
0032 organismID=cellstr(organismID);
0033 fastaFile=cellstr(fastaFile);
0034 modelIDs=cellstr(modelIDs);
0035 refFastaFiles=cellstr(refFastaFiles);
0036 
0037 blastStructure=[];
0038 
0039 <span class="comment">%Get the directory for RAVEN Toolbox. This may not be the easiest or best</span>
0040 <span class="comment">%way to do this</span>
0041 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0042 ravenPath=fileparts(fileparts(ST(I).file));
0043 
0044 <span class="comment">%Construct databases and output file</span>
0045 tmpDB=tempname;
0046 outFile=tempname;
0047 
0048 <span class="comment">%Check for existence of files. If no full path is specified for a file,</span>
0049 <span class="comment">%assume that it is in the current folder</span>
0050 <span class="keyword">if</span> isrow(refFastaFiles)
0051     files=horzcat(fastaFile,refFastaFiles);
0052 <span class="keyword">else</span>
0053     files=vertcat(fastaFile,refFastaFiles);
0054 <span class="keyword">end</span>
0055 
0056 files=checkFileExistence(files,true,false); <span class="comment">%No whitespace allowed</span>
0057 fastaFile = files(1);
0058 refFastaFiles = files(2:end);
0059 
0060 <span class="comment">%Create a database for the new organism and blast each of the refFastaFiles</span>
0061 <span class="comment">%against it</span>
0062 
0063 <span class="keyword">if</span> isunix
0064     <span class="keyword">if</span> ismac
0065         binEnd=<span class="string">'.mac'</span>;
0066     <span class="keyword">else</span>
0067         binEnd=<span class="string">''</span>;
0068     <span class="keyword">end</span>
0069 <span class="keyword">elseif</span> ispc
0070     binEnd=<span class="string">''</span>;
0071 <span class="keyword">else</span>
0072     dispEM(<span class="string">'Unknown OS, exiting.'</span>)
0073     <span class="keyword">return</span>
0074 <span class="keyword">end</span>
0075 
0076 <span class="comment">% Run BLAST multi-threaded to use all logical cores assigned to MATLAB.</span>
0077 cores = evalc(<span class="string">'feature(''numcores'')'</span>);
0078 cores = strsplit(cores, <span class="string">'MATLAB was assigned: '</span>);
0079 cores = regexp(cores{2},<span class="string">'^\d*'</span>,<span class="string">'match'</span>);
0080 cores = cores{1};
0081 
0082 [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'diamond'</span>,[<span class="string">'diamond'</span> binEnd]) <span class="string">'&quot; makedb --in &quot;'</span> fastaFile{1} <span class="string">'&quot; --db &quot;'</span> tmpDB <span class="string">'&quot;'</span>]);
0083 <span class="keyword">if</span> status~=0
0084     EM=[<span class="string">'DIAMOND makedb did not run successfully, error: '</span>, num2str(status)];
0085     dispEM(EM,true);
0086 <span class="keyword">end</span>
0087 
0088 <span class="keyword">for</span> i=1:numel(refFastaFiles)
0089     fprintf([<span class="string">'Running DIAMOND blastp with &quot;'</span> modelIDs{i} <span class="string">'&quot; against &quot;'</span> organismID{1} <span class="string">'&quot;..\n'</span>]);
0090     [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'diamond'</span>,[<span class="string">'diamond'</span> binEnd]) <span class="string">'&quot; blastp --query &quot;'</span> refFastaFiles{i} <span class="string">'&quot; --out &quot;'</span> outFile <span class="string">'_'</span> num2str(i) <span class="string">'&quot; --db &quot;'</span> tmpDB <span class="string">'&quot; --more-sensitive --outfmt 6 qseqid sseqid evalue pident length bitscore ppos --threads '</span> cores ]);
0091     <span class="keyword">if</span> status~=0
0092         EM=[<span class="string">'DIAMOND blastp did not run successfully, error: '</span>, num2str(status)];
0093         dispEM(EM,true);
0094     <span class="keyword">end</span>
0095 <span class="keyword">end</span>
0096 delete([tmpDB <span class="string">'*'</span>]);
0097 
0098 <span class="comment">%Then create a database for each of the reference organisms and blast the</span>
0099 <span class="comment">%new organism against them</span>
0100 <span class="keyword">for</span> i=1:numel(refFastaFiles)
0101     fprintf([<span class="string">'Running DIAMOND blastp with &quot;'</span> organismID{1} <span class="string">'&quot; against &quot;'</span> modelIDs{i} <span class="string">'&quot;..\n'</span>]);
0102     [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'diamond'</span>,[<span class="string">'diamond'</span> binEnd]) <span class="string">'&quot; makedb --in &quot;'</span> refFastaFiles{i} <span class="string">'&quot; --db &quot;'</span> tmpDB <span class="string">'&quot;'</span>]);
0103     <span class="keyword">if</span> status~=0
0104         EM=[<span class="string">'DIAMOND makedb did not run successfully, error: '</span>, num2str(status)];
0105         dispEM(EM,true);
0106     <span class="keyword">end</span>
0107     [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'diamond'</span>,[<span class="string">'diamond'</span> binEnd]) <span class="string">'&quot; blastp --query &quot;'</span> fastaFile{1} <span class="string">'&quot; --out &quot;'</span> outFile <span class="string">'_r'</span> num2str(i) <span class="string">'&quot; --db &quot;'</span> tmpDB <span class="string">'&quot; --more-sensitive --outfmt 6 qseqid sseqid evalue pident length bitscore ppos --threads '</span> cores]);
0108     delete([tmpDB <span class="string">'*'</span>]);
0109     <span class="keyword">if</span> status~=0
0110         EM=[<span class="string">'DIAMOND blastp did not run successfully, error: '</span>, num2str(status)];
0111         dispEM(EM,true);
0112     <span class="keyword">end</span>
0113 <span class="keyword">end</span>
0114 
0115 <span class="comment">%Done with the DIAMOND blastp, do the parsing of the text files</span>
0116 <span class="keyword">for</span> i=1:numel(refFastaFiles)*2
0117     tempStruct=[];
0118     <span class="keyword">if</span> i&lt;=numel(refFastaFiles)
0119         tempStruct.fromId=modelIDs{i};
0120         tempStruct.toId=organismID{1};
0121         A=readtable([outFile <span class="string">'_'</span> num2str(i)],<span class="string">'Delimiter'</span>,<span class="string">'\t'</span>,<span class="string">'Format'</span>,<span class="string">'%s%s%f%f%f%f%f'</span>);
0122     <span class="keyword">else</span>
0123         tempStruct.fromId=organismID{1};
0124         tempStruct.toId=modelIDs{i-numel(refFastaFiles)};
0125         A=readtable([outFile <span class="string">'_r'</span> num2str(i-numel(refFastaFiles))],<span class="string">'Delimiter'</span>,<span class="string">'\t'</span>,<span class="string">'Format'</span>,<span class="string">'%s%s%f%f%f%f%f'</span>);
0126     <span class="keyword">end</span>
0127     tempStruct.fromGenes=A{:,1};
0128     tempStruct.toGenes=A{:,2};
0129     tempStruct.evalue=table2array(A(:,3));
0130     tempStruct.identity=table2array(A(:,4));
0131     tempStruct.aligLen=table2array(A(:,5));
0132     tempStruct.bitscore=table2array(A(:,6));
0133     tempStruct.ppos=table2array(A(:,7));
0134     blastStructure=[blastStructure tempStruct];
0135 <span class="keyword">end</span>
0136 
0137 <span class="comment">%Remove the old tempfiles</span>
0138 delete([outFile <span class="string">'*'</span>]);
0139 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>